### PR TITLE
Record: TMA Megakernel + Triple Loop + Parallel Residuals — val_bpb 1.08480

### PR DESCRIPTION
**Pull request by andrewbaggio1**
# Record: TMA Megakernel + Triple Loop + Parallel Residuals — val_bpb 1.08480

**val_bpb: 1.08480** (5-seed mean, std=0.0007)

| Seed | Steps | **Sliding BPB** | Artifact |
|-|-|-|-|
| 1234 | 4624 | **1.08359** | 15,750,778 |
| 1    | 4613 | **1.08512** | 15,750,312 |
| 42   | 4612 | **1.08530** | 15,751,243 |
| 1337 | 4614 | **1.08448** | 15,747,957 |
| 2025 | 4614 | **1.08553** | 15,752,675 |
| **Mean** | **4615** | **1.08480** | **15,750,593** |

## Summary

This submission combines frontier techniques with a custom Triton TMA megakernel for the MLP forward pass. The fused kernel avoids materializing the 2048-wide intermediate activation, yielding +10.5% training throughput (+127 extra steps in 600s). Together with triple depth recurrence and GPT-J parallel residuals, this achieves 1.08480 BPB.

## What's new

* **Triton TMA Fused MLP Forward (megakernel):** Fuses `fc -> leaky_relu(0.5) -> square` into a single Hopper TMA kernel. Uses `TensorDescriptor` for async global-to-shared transfers, persistent kernel scheduling (one program per SM), interleaved writes, and 128x256x64 block tiling. Avoids materializing ~384MB intermediate per forward pass. Measured +10.5% throughput across all training phases. Falls back to torch.compile for eval.

* **Triple depth recurrence:** Layers 4-5 loop 4 times total (NUM_LOOPS=3, 17 virtual layers from 11 physical). Loop activation at frac=0.35. From [#1420](https://github.com/openai/parameter-golf/pull/1420).

* **Parallel residuals (GPT-J, layers 7-10):** Attention and MLP read from same pre-residual input. From [#1420](https://github.com/openai/parameter-golf/pull/1420).

* **All other techniques from #1394:** SP8192 vocab, GPTQ embeddings (int8), SDClip (k=13.5), MuonEq-R, skip gates, ShuffledSequenceLoader, EMA, brotli.

## Legality

No TTT, no SLOT, no n-gram caching, no multi-pass scoring. Full audit against [#1017](https://github.com/openai/parameter-golf/issues/1017) conditions in the README.

## Reproduction

```bash
rm -f data/manifest.json
MATCHED_FINEWEB_REPO_ID=kevclark/parameter-golf \
python3 data/cached_challenge_fineweb.py --variant sp8192 --train-shards 128

SEED=1234 VOCAB_SIZE=8192 \
torchrun --standalone --nproc_per_node=8 train_gpt.py
```

## Credits

Builds on @clarkkev's [#1394](https://github.com/openai/parameter-golf/pull/1394) and @abaybektursun's [#1420](https://github.com/openai/parameter-golf/pull/1420). Full component lineage in the README.

### CODE REVIEW

**Comment by mohosy** [comment]
tma megakernel is lowkey the most underrated thing in this competition rn, everyones focused on architecture but the speed gains from fused kernels mean more steps in 600s which is free bpb

**Comment by MatoTeziTanka** [comment]
### Community Review — TMA Megakernel + Triple Loop + Parallel Residuals

**BPB:** 1.08480 (5-seed mean, std=0.0007) | **Seeds:** 5 | **Artifact:** 15,750,593 bytes mean (max 15,752,675) | **Vocab:** SP8192 | **Compliance:** PASS

**What this does:** Builds on @clarkkev's #1394 (SP8192, GPTQ embeddings, SDClip, MuonEq-R) and @abaybektursun's #1420 (triple depth recurrence at NUM_LOOPS=3 on layers 4-5, GPT-J style parallel residuals on layers 7-10). The headline contribution is a custom Triton Hopper TMA fused MLP forward kernel (`fc -> leaky_relu(0.5) -> square`) that avoids materializing the 2048-wide intermediate activation; the author measures +10.5% throughput and ~127 extra training steps in the 600s budget.

**What I found in the code (static review at SHA `b27fe936`):**
- **Eval is pure, frozen, single-pass.** `eval_val_sliding` at `train_gpt.py:1201-1272` runs under `torch.inference_mode()`, uses the dequantized `eval_model` loaded with `strict=True` (`:1134`), scores each window on positions `s..wlen` where `s = 0 if ws == 0 else context_size`, and never updates model state between windows. No TTT, no SLOT, no n-gram cache, no multi-pass scoring. The README's audit against the four conditions in #1017 matches the code.
- **Model architecture matches README.** `NUM_LOOPS=3` (`:80`), `parallel_residual_start=7` (`:85`), encoder/decoder loop schedule runs `h.num_loops + 1` times at `:625`. Depth recurrence activates at `frac >= 0.35` (`:1409`). 11 physical layers, 17 virtual with the loop schedule as described.
- **TMA megakernel is well-guarded.** Triton import is in a try/except (`:29-34`); kernel defined at `:314-395` using `TensorDescriptor.from_tensor` for async G→S transfers, persistent kernel over SMs, 128×256×64 block tiling, dual-output writes (forward activation + backward grad factor) in one pass. Falls back to `torch.compile` in eval (`:1209`).
- **Per-seed results table** in `submission.json` matches the README: 5 seeds (1234, 1, 42, 1337, 2025), all artifacts in the 15.747M–15.753M range, every seed under the 16,000,000 byte cap.
- **Artifact accounting:** max per-seed artifact = 15,752,675 bytes (98.45% of budget). `submission.json` reports `bytes_code: 63463`; the train_gpt.py file I fetched at head SHA `b27fe936` is 63,993 bytes on disk (530-byte difference, likely line-ending or whitespace normalization in whichever tool generated the submission.json — does not change the budget verdict since total is 98.45%).
- **Credits section is thorough.** README explicitly attributes #1420 (triple loop, GPT-J parallel residuals, earlier loop activation) to @abaybektursun, #1394 to @clarkkev, #1204 depth recurrence to @msisovic, GPT-J to @kingoflolz, plus full table for SDClip, MuonEq-R, XSA, Partial RoPE, Muon, EMA, Brotli+byte shuffle, sliding window, FA3, etc. The TMA kernel is claimed as author's own work "inspired by #1420" — consistent with the fact that #1420's kernel work is the triple-loop execution, not a fused MLP TMA kernel.
- **Minor README bug:** the reproduction command at the bottom of the README references `train_gpt_mega.py` but the actual file in the PR is `train_gpt.py`. Worth fixing but not blocking.

**TTT compliance verdict:** PASS — no TTT, no eval-time adaptation, no state mutation during eval, `torch.inference_mode()` throughout, single L→R sliding pass with causal context. Per Issue #402 and #677, no score-after-update ordering exists because there is no update at eval at all.

**Gauntlet:** Import OK, Hyperparameters OK (dim=512, layers=11, heads=8, vocab=8192), `ngram_cache=false`. Model-creation step skipped because the harness doesn't know how to pass the `Hyperparameters` arg to `GPT.__init__(h)` — same harness limitation seen on #1420, not a PR defect.

**Questions/flags:**
- None that block merge. One cosmetic ask: reproduction command should be `train_gpt.py`, not `train_gpt_mega.py`.

**Verdict:** LOOKS CLEAN

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** MERGE — beats the current merged SOTA (claimed 0.030 BPB improvement over the merged frontier), all 5 seeds under the 16MB cap, credit lineage is comprehensive and matches what the code actually borrows, eval loop is clearly legal under Issue #1017's four conditions, and the novel contribution (TMA fused MLP forward) is a hardware optimization of a permitted flavor per Issue #1409 (custom kernels allowed).

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU gauntlet: import + hyperparam checks PASS, model-instantiation step skipped (harness doesn't pass the `Hyperparameters` positional arg that `GPT.__init__` requires). AI tooling: review drafted with Claude Code (Sonnet/Opus) using an internal review template; all citations, file paths, and compliance audits were verified against the PR's actual code at SHA `b27fe93647186e7c301928449f2a817d7888e8f3`.*

**Comment by andrewbaggio1** [comment]
thank you but when the reviewers built the merge chain on april 9 they likely skipped it because it didn't fit in the range

**Comment by MatoTeziTanka** [comment]
Makes sense — the April 9 merge chain was focused on a narrower range. The review and MERGE recommendation stand on their own regardless of where it lands in the merge queue. Good luck.
